### ISSUE TITLE
Lightmaps not builded at all...

### ISSUE THREAD

**Issue by Honya2000** [Naxela/The_Lightmapper#129]
Hello,

Today I tried this addon to bake lightmaps in my blender scene and unfortunately it doesn't work.

When I press button "Build Lightmaps" it is doing something for several secs. And after this Lightmaps folder is empty.
Is OpenCV requered for Lightmapper?
I see just one suspicious line in the log about CV2.
If it's required - how to install it? Lightmapper says I have to do this through preferences. But preference window doesn't have such option.

There is log:
Building lightmaps
Warning: 1 x Draw window and swap: 28.1700 ms, average: 28.17000001 ms
Found an object with zero polygons. Skipping object: LM1_Plane.199
PREPARE: ATLAS
PREPARE: OBJECTS
Config mesh catch omitted: REMEMBER TO SET IT BACK NAXELA
Beginning build
Managing build
GPU Encoding
CV2 not found - Ignoring postpacking
Lightmap building finished
2021-11-07 10:59:30.599848: Logger started.
2021-11-07 11:00:17.900307: Logger started.
2021-11-07 11:00:17.900307: Preparing build
2021-11-07 11:00:26.882333: ['0.0 hours 0.0 mins 9.0 seconds']
2021-11-07 11:00:26.883333: Lightmap building finished

**Comment by Honya2000** [comment]
I was able to solve OpenCV issue by manual installing and copying it to blender python lib folder.
But unfortunately it didn't solve my issues with Lightmap generation.

My scene is quite complex. But currently i'm experimenting with single floor object.
After baking the floor is always black. Though lightmap has some ligting in it. Looks like UV mapping issue.
I tried almost all possible options regarding Unwrapping, Packing, etc. Sometime python throws exceptions on some options. But In 4 hours I didn't manage to make it working. The floor remains black.

Will try to experiment with other objects...

**Comment by Honya2000** [comment]
Ok, I checked with external walls object. And seems like lightmapping is working on this object.
But problem is,
1. floor is most important object for light-mapping.
2. as I said the scene is quite complex. I'm almost 100% sure if it doesn't work for floor, it will fail for some other objects in the scene as well.  

Any suggestions how to make floor lightmaps working?

P.S. I can share my scene via google drive if needed.

**Comment by Honya2000** [comment]
I gave up. Nothing works for me.
Some settings give me black lightmap on floor. Some puple. But no any setting allowed me to see any shadows on the floor.  So the floor is blocker definitely.

On some objects lightmap work, but seems like it blured to mess. So no any details from trees leaves are visible (Just solid shadow). Even without denoiser (Filter is disabled). Resolution is 2048. Should be enough...

**Comment by Naxela** [comment]
Hi, could you share the scene? If they're purple it's mostly due to having had other errors breaking things, although the black faces can have a lot of reasons, ranging from normals to UV usage

**Comment by Honya2000** [comment]
I'm uploading the scene to google drive. Doubt the reason is relating to normals. The same object works fine in another light maps baker I used before.
Just wanted to try this one because of HDR light maps support (and probably better atlas management)... But atlas thing i will test later after floor fix.
Just tried once to use Atlas prepack (instead of Smart project) and got python exception ))

**Comment by Honya2000** [comment]
There is link to blend scene.
https://drive.google.com/file/d/1PabbDo9lD8QwXXAbOjDbdB1jFuy-tczw/view?usp=sharing

Just enable Light Mapper for object: Terrain.012.

There are 3 issues so far:
1. Terrain object is always black after light maps generation.
2. Cannot get detailed smooth shadows. For example from trees leaves. Shadows looks like solid.
3. On grass material (which is part of Terrain object) shadows have some bright silhouettes. Not sure is it HDR encoding related (RGBD) or something else. But I couldn't make lightmapper working without encoding. Always get purple surface.
 
Thank you!

**Comment by Naxela** [comment]
Hi,

I tested a bit further with your model, but to first reply to your other questions:

- OpenCV isn't required for the lightmapper, but it can help smoothen your lightmaps.
It's installed by opening Blender as an administrator, and clicking Install OpenCV:

![image](https://user-images.githubusercontent.com/5429817/140724312-c8efa9f8-89f1-4509-85fb-14be95134c67.png)

---

As for the model:

I tried moving your floor (terrain.012) to a separate scene, and lightmapping it. It seems
by default the UV utilization is poor due to the object having a lot of faces and too much margin.
On my end, setting the lightmap resolution to at least 512 and using 0.01 instead of 0.1 as unwrap
margin, it seemed to do the trick:

_Before:_

![image](https://user-images.githubusercontent.com/5429817/140723386-326612fa-8398-4203-ad42-41354127571a.png)

After:

![image](https://user-images.githubusercontent.com/5429817/140723906-7ec72f3a-5e79-4708-840d-d5d76032d966.png)

By default, once your model has already been unwrapped, you'll either need to manually delete
the "UV_Unwrap" UV channel to try other UV packing algorithms. Or use one of these methods:

- Toggle Remove Lightmap UV, it will remove all the lightmap UV channels upon build
- Select the object(s) and go to Utilities, set to selection and click Remove Lightmap UV. 

As for the shadow from the leaves, do you get them even when you render with Cycles? I didn't
get them by default on a regular render:

![image](https://user-images.githubusercontent.com/5429817/140724498-265a7501-f5c1-4686-a515-c39c3c12ab99.png)

---

Not getting the lightmapper to work without encoding sounds strange. By default it should work,
but it could be a video memory issue if you bake all your objects (1600'ish?), with about 256 to
512 resolution, that could add up to many gigabytes of HDR files which the VRAM might not
be too happy about - could explain why encoding works. Does it work if you're just lightmapping
one object? - A possible fix for this might be to use Atlases

As for the encoding, these are mainly targeted for game engines, so once encoded they need to
be decoded as well, which is why RGBD might give some strange highlights.

**Comment by Honya2000** [comment]
Thank you for investigating!

Today I tried "Lightmap" unwrapper instead of Smart Project, and it works to some degree.
It's not fully satisfied me because there are lot of stair artefacts on the polygon edges of the floor. Like lightmap UV's are not  stick accross polygons and not share vertices. Will post screen later.
Though I used high lightmap settings, 4k resolution and supersampling and it took over one hour to generate lightmap just for one object..

Strange that I never see this artefacts with my previous much simpler lightmapper, which doesn't have all this advanced settings.

Another problem:
When I export model after baking to GLB - it exports lightmap texture in base color slot instead of albedo.
So I don't have albedo anymore.

Will debug GLB loading a little bit. Probably it exports albedo in diffuse slot and lightmap in base color slot.

**Comment by Honya2000** [comment]
Unfortunately setting Unwrap Margin 0.01 didn't help if i choose "Smart Project" unwrap method.
Regarding clearing of UV's.
First. By default terrain object doesn't have second UV set. So I assume lightmapper created LightMap UV itself.
Second - i tried each option related to UV: "Remove Lightmap UV: button. Tick "Remove Lightmap UV" checkbox.
Didn't help.

Not sure which options remain to check. Unwrapping method seems like changing something. But atlas methods are throwing exception. 
So only Lihtmap method kinda works. But will need to somehow fix this stair artefacts...

**Comment by Naxela** [comment]
Strange, I'll look into it.

The problem with the "lightmap" algorithm is that it's prone to stretching and triangulation which causes a whole bunch of weird issues on surfaces that are connected and/or not entirely flat.

In the meantime, I can recommend install this addon: https://github.com/mattedicksoncom/blender-xatlas - It's possibly the best free automatic unwrapping solution for Blender, and there's native support for it in the addon.

**Comment by Honya2000** [comment]
So which is workflow with xatlas?
Should I first unwrap UV's with this addon and then run lightmapper?
Or lightmapper automatically will utilize this addon during UV's generation ?

**Comment by Naxela** [comment]
Once it's installed, it should be detected by the lightmapper and available as an unwrapping option (like smart project, lightmap, etc.) and just use the lightmapper as usual

**Comment by Honya2000** [comment]
Also didn't understand how to work with Atlas Group (Prepack) unwrapping method.
After baking Shader node structure was not changed. So lightmap was generated but not applyed for some reason.
I expect there should be some materials merge separate button, which I can press to merge materials which utilize the same atlas. But there is no such button.
So how to apply atlas to the materials, so I can later export it to GLB ?

**Comment by Honya2000** [comment]
Unfortunately doesn't work with XAtlas:
Got exception:

Error: Python: Traceback (most recent call last):
  File "C:\Users\BLA\AppData\Roaming\Blender Foundation\Blender\2.93\scripts\addons\The_Lightmapper-master\addon\operators\tlm.py", line 52, in invoke
    build.prepare_build(self, False)
  File "C:\Users\BLA\AppData\Roaming\Blender Foundation\Blender\2.93\scripts\addons\The_Lightmapper-master\addon\utility\build.py", line 95, in prepare_build
    prepare.init(self, previous_settings)
  File "C:\Users\BLA\AppData\Roaming\Blender Foundation\Blender\2.93\scripts\addons\The_Lightmapper-master\addon\utility\cycles\prepare.py", line 24, in init
    configure_meshes(self)
  File "C:\Users\BLA\AppData\Roaming\Blender Foundation\Blender\2.93\scripts\addons\The_Lightmapper-master\addon\utility\cycles\prepare.py", line 425, in configure_meshes
    Unwrap_Lightmap_Group_Xatlas_2_headless_call(obj)
  File "C:\Users\BLA\AppData\Roaming\Blender Foundation\Blender\2.93\scripts\addons\The_Lightmapper-master\addon\utility\utility.py", line 495, in Unwrap_Lightmap_Group_Xatlas_2_headless_call
    stdin.write(value)
BrokenPipeError: [Errno 32] Broken pipe

**Comment by Naxela** [comment]
Strange, did you use version 0.08 from the releases page?: https://github.com/mattedicksoncom/blender-xatlas/releases

Does xatlas work if you use it alone (from the sidemenu)?

**Comment by Honya2000** [comment]
Actually i checked. Xatlas was installed before in Users/AppData/Roaming/Blender/... folder folder.
And today i copied it over to original program files/Blender addons folder.

Probably there was conflict.
Will try to delete it from Program Files.

**Comment by Honya2000** [comment]
Ok, there are 2 ways to install XAtlas.
1. Using zip which i cloned from git url you gave me.
2. To copy addons folder from this url durectly in Blender addons.

First method doesn't work for me, it installs entire content of zip to Users blender addons path. But Blender doesn't recognize it. I guess because zip folder structure is not suitable for blender.

If I copy it to addons folder myself - blender recognizes it, so i can enable and see it in side menu.
But in this case lightmapper throws exception.

Will try to copy it manually to user blender addon folder instead of Program Files.

**Comment by Honya2000** [comment]
This time it worked without exceptions, but produce absolutely random UV's or lightmap.

https://photos.app.goo.gl/nqm1hALH1pw551Nv8

**Comment by Honya2000** [comment]
Is ther way to Unwrap UV's by ranning standalone XAtlas addon. And then reuse UV's in Lightmapper ?

**Comment by Naxela** [comment]
That seems odd, I got this just by setting it to Xatlas and 1024 resolution. I also applied the rotation (I saw that it was at 180 degrees, not sure if that makes a difference):

![image](https://user-images.githubusercontent.com/5429817/140749268-b2cf512a-7815-4f95-aac8-7a14715508f0.png)

![image](https://user-images.githubusercontent.com/5429817/140749334-36938372-5647-4209-accb-ca20d54e6c2d.png)

I probably should start doing some video tutorials soon

**Comment by Naxela** [comment]
I think Blender_Xatlas should create a UVMap_Lightmap channel by default, if you use the addon itself:

![image](https://user-images.githubusercontent.com/5429817/140749568-4bf6cb38-8f77-4f69-93bb-8b1ef62e5135.png)

**Comment by Honya2000** [comment]
Then I don't know what is going on.
I tried 1024 resolution for XAtlas, and 2048.
Also tried to run XAtlas standalone prior to Lightmapper.
It's always produce random pixels.

And not sure where did you get 180 degrees. I checked object transform. And it was identity (0 degrees for all axis).

**Comment by Honya2000** [comment]
And again.
Weird thing that this is not happening on another object: Exterior walls.
At least lightmap image doesn't look random.

I don't understand what is wrong with this floor!

**Comment by Naxela** [comment]
I got that value from the file you uploaded:

![image](https://user-images.githubusercontent.com/5429817/140757919-fa709d30-7cbf-4a58-9275-331ea4daeaae.png)

I can try make a video/screencast when I get home, showing what I do to fix it

Also, I think using the master branch should be fine

**Comment by Honya2000** [comment]
Regarding rotation 180.
Yes. in original file it is 180 deg.
But it becomes 0 as soon as I press button "Enable for set" in utilities section of the tool.

But rotation doesn't change anything.
I just tried right now with original file after rotating terrain to 0 degrees.
It still produces random image with XAtlas enabled.

**Comment by Honya2000** [comment]
Tried to change resolution in XAtlas tool to 1024, tried to apply transform, tried without apply, any combinations of those settings.
Result is the same random.

**Comment by Honya2000** [comment]
So XAtlas is not an option for me unfortunately. Spend entire day on it and could't make it working.

Tried unwrapping method Atlas Group. But it is very unstable.
In about 90% of cases i get exception:
Error: Python: Traceback (most recent call last):
  File "C:\Users\BLA\AppData\Roaming\Blender Foundation\Blender\2.93\scripts\addons\The_Lightmapper-master\addon\operators\tlm.py", line 52, in invoke
    build.prepare_build(self, False)
  File "C:\Users\BLA\AppData\Roaming\Blender Foundation\Blender\2.93\scripts\addons\The_Lightmapper-master\addon\utility\build.py", line 95, in prepare_build
    prepare.init(self, previous_settings)
  File "C:\Users\BLA\AppData\Roaming\Blender Foundation\Blender\2.93\scripts\addons\The_Lightmapper-master\addon\utility\cycles\prepare.py", line 24, in init
    configure_meshes(self)
  File "C:\Users\BLA\AppData\Roaming\Blender Foundation\Blender\2.93\scripts\addons\The_Lightmapper-master\addon\utility\cycles\prepare.py", line 287, in configure_meshes
    bpy.context.scene.UVPackerProps.uvp_padding = atlasgroup.tlm_uv_packer_padding
AttributeError: 'Scene' object has no attribute 'UVPackerProps'

When there is no exception (it's quite difficult to predict when it will crash, in which situation) - it generates lightmap image but doesn't apply it to shaders.

**Comment by Honya2000** [comment]
Btw just tried to run standalone XAtlas addon to generate UV's prior to LM generation.
And after this generated light map using Smart project UV's unwrapping.
Sure UV's was not generated during lightmapper pass, because this was already pregenerated by XAtlas.

So as expected It generated mess instead of image...
Definitely issue with XAtlas.

**Comment by Honya2000** [comment]
Tried XAtlas with another lightmapper. Also got mess. So don't think it is the best one!

**Comment by Naxela** [comment]
Just got home, I'll share a screencapture video of the steps I take to get the results

**Comment by Honya2000** [comment]
Ok, I did exactly all the steps, you provided.
The same mess after baking.
The issue is definitely in XAtlas.

Are you sure you use latest XAtlas?
Also i see your Lightmapper addon is a little bit different.
For example in Lightmapper section of object properties you have check box "Use UV Packer".
I don't have this option.

But most important is the log!
You showed the end of your log. And it differs from my log.

There is mine:
`182, 14180], [14023, 14024, 14025], [14031, 14032, 14033], [14039, 14042, 14040], [13020, 13019, 13021], [13018, 13024, 13019], [32, 33, 34], [2549, 2550, 2548], [2395, 2394, 2396], [13062, 13063, 13064], [14207, 14210, 14208]])]
Applying the UVs----------------------------------------
Finished Xatlas----------------------------------------
Config mesh catch omitted: REMEMBER TO SET IT BACK NAXELA
Beginning build
Baking 0/1 (0.0%) : Terrain.012
Elapsed time: 5.66s | ETA remaining: 5.66s
Baking combined: Direct + Indirect
Info: Baking map saved to internal image, save it externally or pack it
K:\Projects\_Models\Lightmaps\AtlasGroup_baked.hdr
K:\Projects\_Models\Lightmaps\Terrain.012_baked.hdr
Managing build
CHECKING FOR REPART
Lightmap building finished
2021-11-09 07:58:52.397390: Logger started.
2021-11-09 08:19:00.866787: Logger started.
2021-11-09 08:19:00.867787: Preparing build
2021-11-09 08:21:07.600773: ['0.0 hours 2.0 mins 6.7 seconds']
2021-11-09 08:21:07.603073: Lightmap building finished`

First of all why triangle indices generated by XAtlas are different if model is the same ?
Second: in your log:
`PREPARE: ATLAS
PREPARE: OBJECTS
Preprocessing material for: Terrain.012
Managing layer for Obj: Terrain.012`

This lines printed after XAtlas processing.
In my version - before XAtlas.

So maybe issue not in XAtlas buf in version of Lightmapper i'm using.

**Comment by Honya2000** [comment]
Before I used GLB Texture Tools for light maps generation.
It doesn't have many options for UV's Unwrapping. Just one check box "Unwrap".
Analyzed a little bit the code. They use Smart Project for unwrapping.
Question is why I never had this issues with generating lightmaps using their tool?

For me Smart Project unwrapping option in TLM just doesn't work. It generates black terrain.

**Comment by Honya2000** [comment]
My target is:
1. Select group of objects
2. Assign Atlas name to this group (or atlas id, whatever).
3. Press bake lightmap.
4. Repeat above 3 steps for another group.
4. Export to GLB.

Unfortunately so far all above steps are failed for me using TLM.

I don't know how to assign atlas to the selected objects.
Also many questions to GLB export.

Potential of this project is great. It looks awesome on the paper and video. HDR, many options, denoising, filtering, etc.
But all this useless if just simple bake for one object produces wrong result.

And btw pretransforming of all objects to global space also not really good idea.

**Comment by Naxela** [comment]
Hi,

Yes, it's important to unhide the hidden elements in edit mode, as the lightmapper uses the available selectable faces to unwrap, otherwise it unwraps only the visible ones which becomes a mess in object mode. - Basically just go to edit mode, press ALT + H.

As for Xatlas, I use version 0.08 x64

As for the different version, it the UVpacker option should be in the master - although it's just stubs code for now. 

I think the GLB tools might unhide all elements before baking, I should probably add that as well. 

As for pretransforming (if you mean applying scale and rotation), the internal UV unwrapping tools are dependent on this unless you want to have skewed/rotated - See this issue: https://github.com/Naxela/The_Lightmapper/issues/3 - Although this shouldn't be the case if you manually UV your objects.

**Comment by Honya2000** [comment]
Alt-H did the trick!
After 3 days I got something. Will continue experimenting ))
So can I select objects group, Unhide elements, apply transforms, then generate UV's using standalone XAtlas, save model, and only after this start Lightmapper ?
If i specify for this group texture res 2048 in XAtlas - will Lightmapper put this group to one atlas?

**Comment by Naxela** [comment]
That's good to hear, I added a separate issue for my to do list related to this: https://github.com/Naxela/The_Lightmapper/issues/130

Yes, you can do that manually although it should in theory already do all of that by default (except unhide elements) when Xatlas is selected. - And yes, but with prepack atlas it will override your existing and use the one you've selected in the atlas section. Postpack atlas will take your existing object UV's and pack them together into small squares (bins)

**Comment by Honya2000** [comment]
Regarding prepack. As I mentioned above I didn't understand how to use it.
It generates lightmap texture. But doesn't apply it to materials. 
Manually adding nodes is not an option...

So i don't know how to preview it in blender or how to export it.

**Comment by Naxela** [comment]
Depends a bit on what you want to achieve - Are you thinking you want the objects to share the same material and lightmap? Or just the same lightmap across multiple materials - The lightmapper only does the latter

Usually, it should apply the lightmap to the material at the end, but I'll double check if I've added some bugs lately

**Comment by Honya2000** [comment]
Most important is to utilize atlas. So i need to share one texture for selected objects 
Sharing of materials would be bonus but not so important. I can do this in my engine during GLB loading automatically.
So far I couldn't achive sharing of texture for 2 selected objects with XAtlas.

So I selected 2 couch objects. Unshared mesh (Made single user).
Selected for both XAtlas 1024. And baked. It generated 2 textures for each object. But I want all textures packed to atlas.
I also tried to unwarap with standalone XAtlas. Still generates 2 textures.

 As far as I understood with XAtlas Unwrapping this is not achievable? ))
So how to pack all selected objects to one atlas (with XAtlas or without) ?

**Comment by Honya2000** [comment]
1. texture resolution should be specified for atlas.
2. unwrapper tries to unwrap all selected objects by fitting them to the atlas with specific resolution. So texel density depends on number of objects selected and atlas resolution.

**Comment by Naxela** [comment]
The process of building an atlas prepack lightmap is as follows:

Scroll down to additional
Make an atlas (click +)
Set unwrapping to Xatlas
Set atlas lightmap resolution
Go to each object and do this for each:
- Enable lightmapping
- As unwrapping method, select Atlas Group (Prepack)
- In the atlas group, select your created atlas

Build

Each object has their respective materials, but share the same lightmap (atlas lightmap)

![image](https://user-images.githubusercontent.com/5429817/140902306-eea7c6c7-e984-4161-a431-a25ce2e59747.png)

![image](https://user-images.githubusercontent.com/5429817/140902323-e9946877-a1cf-4edc-8e9e-6e98cc747b88.png)

![image](https://user-images.githubusercontent.com/5429817/140902345-5dd5404f-5d2f-4c1c-957e-4718102e59b0.png)

![image](https://user-images.githubusercontent.com/5429817/140902365-f1dd6747-a8c5-4e77-a4bf-fb21c4993dbe.png)

---

Originally, TLM wasn't made for game engines either, but just for archviz - I have been working on something like that, although it's still far from finished, but it's a binding for the Texel Density addon: https://github.com/mrven/Blender-Texel-Density-Checker. Also I didn't know CryEngine had lightmapping

**Comment by Naxela** [comment]
As for texel density in terms of atlases, if you use postpack, it doesn't "cram" objects together, it keeps the object's individual resolution and stores it as a bin in a larger atlas lightmap. That's shown as the utilization ratio for object resolutions compared to the lightmap resolution:

![image](https://user-images.githubusercontent.com/5429817/140904069-815182f0-8982-4e2c-a142-26a56b8d456a.png)

**Comment by Honya2000** [comment]
CryEngine has Dot3 lightmapper (with dominant light encoding). But currently it used only for mobile targets.

**Comment by Honya2000** [comment]
Right.
Was able to bake 2 objects to large atlas with post bake.
Only issue is... Why it looks so bad.
With GLB texture tools I baked about 100 objects in outdoor area to one atlas 2048. And quaity was better.
Here i selected 2 couch objects. Each has 512 resolution. And Post-packed it to atlas 1024.
Result is below...

![image](https://user-images.githubusercontent.com/34769255/140905776-f25b7221-ba9b-439f-968f-e67e2c037320.png)

**Comment by Honya2000** [comment]
Btw will try different LM quality settings...

**Comment by Naxela** [comment]
I had a quick try at it - Not sure what causes it to go dark, but the shadows looks similar to the render:

Lightmap (2x 256 resolution to 512 postpacked atlas; Interior Previz quality + denoise):
![image](https://user-images.githubusercontent.com/5429817/140908545-9be5a09e-fa9e-444d-9f25-f16c39ea2671.png)

Render:
![image](https://user-images.githubusercontent.com/5429817/140908567-ef349c76-f5fd-45d4-90a6-852f38cf2cdb.png)

**Comment by Naxela** [comment]
Btw, any specific reason for not using GLB tools instead of The lightmapper?

**Comment by Honya2000** [comment]
Hmm, on your screen it looks much much better.

GLB texture tools have other problems, not related to Unwrapping ))
First only SDR. But i need high range. Second, i didn't manage to fix broken objects. So if lightmap is broken on some objects, i cannot reassign atlas and rebuild it. Obly building entire group from scratch. Which is very time consuming.

**Comment by Naxela** [comment]
Did you remember to make it single-objects with applied scale and rotation? - Also, I toggled off sheen and specularity in the material (for the Eevee viewport)

**Comment by Honya2000** [comment]
I set LM quality to high and it is still building already 20 mins for 2 objects.
I even cannot imagine how long it will build with production quality for entire scene! )

**Comment by Honya2000** [comment]
Ok, in one hour i got 2 objects generated.
Will check now how exporting to GLB works...

And of course packing to atlases... Is still issue. I cannot just select random number of objects and bake them to one atlas.
With post packing I have to make sure the objects I selected will fit the atlas I choose. Such workflow is kinda pain if i have thousands of objects.
Will have to build without atlases first. And later will optimize...

**Comment by Naxela** [comment]
In the future, I can probably make some kind of automatic distribution tool, that can make atlases based on the amount of selected objects and their resolution

**Comment by Naxela** [comment]
By the way, if you find out how GLB works in terms of exporting and find some setups that works better, I'd be happy to make a utility that eases the process if it's just a question of re-arranging some materials or similar

**Comment by Honya2000** [comment]
I don't see anything special in GLB Tex Tools nodes after bake.
Lightmap multiplies to Albedo and assigned to color slot of Principled shader.

But for some reason their setup exported fine to GLB. Albedo and Lightmap are in separate slot.

LTM nodes setup somehow not recognized by GLB exporter and exports Lightmap in Base Color Slot. So no Albedo anymore. Just the lightmap.

This is nodes setup generated by GLB:

![image](https://user-images.githubusercontent.com/34769255/140934058-ce96fe83-f2ba-437e-9a3d-21c4633aa2a2.png)

**Comment by Honya2000** [comment]
Just found here. Lightmap can be stored in AO slot. This would be perfect solution:

https://docs.blender.org/manual/en/2.80/addons/io_scene_gltf2.html

Baked Ambient Occlusion
glTF is capable of storing a baked ambient occlusion map. Currently there is no arrangement of nodes that causes Blender to use such a map in exactly the same way as intended in glTF. However, if the exporter finds a custom node group by the name of glTF Settings, and finds an input named Occlusion on that node group, it will look for an Image Texture attached there to use as the occlusion map in glTF.

**Comment by Honya2000** [comment]
Experimented with GLB exporter.

There is correct nodes setup where GLB exporter exports everything right.

Very important:
Lightmap should be connected to second input of Multiplier. If it connects to first one - LM will be always exported in Diffuse slot instead of Diffuse texture.

And also lightmap texture should be connected to glTF Settings node (Occlusion input).
Otherwise GLB exporter ignores it.

![image](https://user-images.githubusercontent.com/34769255/140998327-6ae0f4ba-36b1-4e2c-8c6c-046b3a7f09f2.png)

**Comment by Honya2000** [comment]
Hi, I have question about incremental lightmaps generation.
If I already generated LM for one group of objects and want to build another group. How can I do this?

Today I wanted to build LM for external walls (walls where selected), while terrain was generated yesterday.
And it generated both terrain and walls for me. I expected it should generate LM for selected objects only? Or?

If it generates all objects which have enabled lightmap - then this is problematic, i cannot manually disable light-mapping for all objects which already generated prior to generating new group.

**Comment by Naxela** [comment]
If it would help, making a script to switch the order would be fairly simple. As for the occlusion node, is that simply a node named Occlusion?

As for incremental lightmap generation, the way it's currently intended is to have selections based on collections, and simply add/remove the enabled lightmapping based on this. Although I realize that it can be quite tedious, so I've been thinking of adding Lightmap Groups, which can be numbered, and then those groups can be toggled for baking, perhaps with separate output folders too

**Comment by Honya2000** [comment]
I'm not sure how glTF Settings node is created. GLTF documentation says: this is custom node with name glTF Settings.
I can check GLB Texture Tools how it creates it in python code.
Probably will modify TLM code myself...

**Comment by Honya2000** [comment]
Regarding Incremental LM generation. I would prefer to generate just selected objects.
Splitting by groups is probably ok on initial setup. But what if I need to fix specific objects in the group after baking? Ungroup and Re-Group?

GLB Texture Tools has affect settings combo box: Affects Selected / Affects Scene, Affects Visible, etc.

I always used Affects Selected. 
Selecting of groups is very flexible using entire collections or name patterns, or just custom manual selection, whatewer. So this option works in almost all scenarious.

**Comment by Honya2000** [comment]
def add_gltf_settings_node(self, material):
        nodes = material.node_tree.nodes
         # create group data
        gltf_settings = bpy.data.node_groups.get('glTF Settings')
        if gltf_settings is None:
            bpy.data.node_groups.new('glTF Settings', 'ShaderNodeTree')
        
        # add group to node tree
        gltf_settings_node = nodes.get('glTF Settings')
        if gltf_settings_node is None:
            gltf_settings_node = nodes.new('ShaderNodeGroup')
            gltf_settings_node.name = 'glTF Settings'
            gltf_settings_node.node_tree = bpy.data.node_groups['glTF Settings']

        # create group inputs
        if gltf_settings_node.inputs.get('Occlusion') is None:
            gltf_settings_node.inputs.new('NodeSocketFloat','Occlusion')

        return gltf_settings_node

**Comment by Honya2000** [comment]
Ok, i modified the script. Now it setups all nodes right.
But I found very serious issue.

If I want to generate LM for new object (single) after generating of other objects currently I have to disable LM for previous object (s).
But in this case Lightmapper just clears lightmap for those previous generated objects.
So i cannot generate specific objects incrementaly at all. Just generating of entire scene. Which is crazy, because each generation will take several days.
Without incremental generation all this is useless.

Will try to fix this myself.

**Comment by Fairplex90** [comment]
>Without incremental generation all this is useless.

You should look for BakeTool, you can setup "jobs" to bake direct and indirect lighting, you can toggle the jobs as you wish. I agree that Lightmapper totally needs that "Bake On Demand" feature:

![Screenshot_5](https://user-images.githubusercontent.com/31825109/141404146-496aaad6-0781-4ec3-823a-0ab954bf2bc9.png)

**Comment by Honya2000** [comment]
I didn't know about BakeTool. Will check.

Yesterday I made all necessary changes in TLM scripts.
Now it fully support Building and Cleaning on selections. And export everything correctly to GLB.
Baked half of scene in several steps already. No issues so far.
